net: coap_client: Various fixes for CoAP client

SeppoTakalo opened the pull request:
This PR is a list for fixes and tests for CoAP client and one new API for cancelling a request.
There should be no behavioral changes, other than correctly responding to some error cases.
It might be easier to review each commit separately, if you need details.

## net: lib: coap_client: Add API to cancel specific request

Add a new API to cancel just one, or mathing requests,
instead of cancelling all ongoing requests.

## net: lib: coap_client: Stop polling on unneeded sockets

poll() only for sockets that have traffic ongoing or have some lifetime
left.
On socket failures during a poll(), stop listening for the socket.
Application can recover by reconnecting the socket. It will be polled when application successfully sends next query.
 
## net: lib: coap_client: All error cases should be reported to callback

When the client fails when parsing the response and we stop proceeding,
we should report that to the application.
  
## net: lib: coap_client: Drop duplicate MID only after responding with Ack

Even if we receive duplicate confirmable message, we should still
respond with the Ack. Just don't deliver the second callback.
As per RFC: https://datatracker.ietf.org/doc/html/rfc7252#section-4.5

## net: lib: coap_client: Drop duplicate responses

When response is already handled, don't forward anymore responses
to the client application.
   
## net: lib: coap_client: Don't match zero length tokens

If our internal structure is cleared, don't match tokens.

## net: lib: coap_client: Don't clear internal structures on response

When response is received and handled, don't just clear the structure
but instead mark it as ongoing=false.
So if we later on receive a duplicate response for it, we can still
respond with Ack or Rst.

This is achieved by using release_internal_request() when we don't
expect any response for it and reset_internal_request() when we really
fill up a new request.

## Various test improvements

* Add testcase for receiving RST
* Add testcase for observation
* Test for operating on socket while another fails
* Add testcase for Ack failure
* Improve testcases for testing client's retry behaviour
* Add test when separate response is lost
* Add test where we receive same UDP packet twice
* Improve socket stubs

SeppoTakalo committed:
net: lib: coap_client: Don't clear internal structures on response

When response is received and handled, don't just clear the structure
but instead mark it as ongoing=false.
So if we later on receive a duplicate response for it, we can still
respond with Ack or Rst.

This is achieved by using release_internal_request() when we don't
expect any response for it and reset_internal_request() when we really
fill up a new request.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Parse incoming MID only once

Incomming Message-ID is already parsed, use it as a parameter to
get_request_with_mid().

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Drop duplicate responses

When response is already handled, don't forward anymore responses
to the client application.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Improve socket stubs

* Use sys_rand_get() and seed the CoAP library, so our MIDs are
  random.
* Set socket events per socket, so we don't accidentally receive
  on wrong socket
* Reply with correct tokens

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Add test for duplicate response

Add test where we receive same UDP packet twice.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Add test when separate response is lost

Test a scenario where Ack is received but the actual response
is not coming.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Return -errno from send_request()

Return the -errno when zsock_sendto() or zsock_recvfrom() fails, so
rest of the code can deal with return values, instead of separately
comparing errno and return value.


Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Drop duplicate MID only after responding with Ack

Even if we receive duplicate confirmable message, we should still
respond with the Ack. Just don't deliver the second callback.

This is achieved by moving the MID deduplication to after Ack handling.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: All error cases should be reported to callback

When the client fails when parsing the response and we stop proceeding,
we should report that to the application.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Don't match zero length tokens

If our internal structure is cleared, don't match tokens.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Improve retry testcases

Add testcases for testing client's retry behaviour.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Add testcase for Ack failure

Add testcase where sending Ack to incomming Confirmable message fails.
This should be reported to application as now the server is unaware
that transmission have succeeded, so we cannot thread it as success.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
net: lib: coap_client: Stop polling on unneeded sockets

poll() only for sockets that have traffic ongoing or have some lifetime
left.
On socket failures during a poll(), stop listening for the socket.
Application can recover by reconnecting the socket.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Test for operating on socket while another fails

CoAP client should be able to push data through functioning socket
while another sockets is failing or reporting poll() errors.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Add testcase for observation

Add test for ongoing observation and cancellation.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo committed:
tests: coap_client: Add testcase for receiving RST

When server responds with CoAP RESET, we should inform the
client and stop the request.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

rlubos commented on tests/net/lib/coap_client/src/main.c:
```diff
@@ -659,6 +674,36 @@ ZTEST(coap_client, test_separate_response)
+	ret = coap_client_req(&client, 0, &address, &client_request, NULL);
+	zassert_true(ret >= 0, "Sending request failed, %d", ret);
+
+	zassert_ok(k_sem_take(&sem, K_MSEC(COAP_SEPARATE_TIMEOUT * 100)));
```
6000 * 100 msec timeout (10 minutes), was that intended? Seems quite lengthy for a test case.

SeppoTakalo replied:
Oops.. left over from debug session. It should be just `COAP_SEPARATE_TIMEOUT`, will fix.

SeppoTakalo committed:
net: lib: coap_client: Add API to cancel specific request

Add a new API to cancel just one, or mathing requests,
instead of cancelling all ongoing requests.

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

jeffwelder-ellenbytech commented:
https://github.com/SeppoTakalo/zephyr/blob/b9719af106f3bb09333e8178bd4b4a857f4c49f5/subsys/net/lib/coap/coap_client.c#L419-L442

@SeppoTakalo inside `coap_client_req()` for a non-confirmable request do you still have to `coap_client_schedule_poll()`, and also now free the request on successful send with `reset_internal_request()`?

I was testing your code, and adding some ping logic (based on our discord conversation), and ran into problems with non-confirmable messages clogging up `client->requests` with `request_ongoing = true` so new requests wouldn't get processed.

I was able to fix this condition by moving `coap_client_schedule_poll()` into the check for Confirmable, and also call `reset_internal_request()` when send_request was successful for a Confirmable message.

Happy to discuss further, maybe I have some understanding wrong as I just started diving into this code.
